File: ts/budget.ts
/**
 * Budget — a ledger of turns, time, and tokens against caps. Pure: it never
 * reads a clock or counts a token; the caller charges what it measured.
 *
 * Lifted from a chat agent's per-question budget (a turn cap, a wall-clock
 * cap, an input-plus-output token cap), where exhaustion ends the loop with
 * "the findings so far" instead of an error. This part keeps the accounting
 * and the exhaustion verdict; what to do when exhausted stays with the caller.
 *
 * Departures from the source: every cap is optional (null: no cap); a budget
 * is exhausted when the amount used reaches its cap, for all three resources
 * alike (the source compared turns with >= but time and tokens with >); the
 * ledger holds 64-bit integers and its additions saturate at 2^63 - 1.
 *
 * Caps and amounts are non-negative integers up to 2^63 - 1, as numbers or
 * bigints, in maps holding no other field. Anything else is out of contract,
 * like a string handed to the breaker: these are the caller's own literals,
 * not data a verdict must speak to, so they throw.
 */
import { checkFields, integer } from "./contract.ts";
import { show } from "./messages.ts";
import { verdict, type Verdict } from "./verdict.ts";

/** 2^63 - 1: the largest value every port's ledger holds. */
const LEDGER_MAX = 9223372036854775807n;

export interface BudgetCaps {
  /** Turns (calls, iterations) allowed; null or absent for no cap. */
  max_turns?: number | bigint | null;
  /** Milliseconds allowed; null or absent for no cap. */
  time_budget_ms?: number | bigint | null;
  /** Tokens allowed, counted however the caller counts them; null or absent for no cap. */
  token_budget?: number | bigint | null;
}

/** What was used; an amount that is null or not given is nothing used. */
export interface Charge {
  turns?: number | bigint | null;
  ms?: number | bigint | null;
  tokens?: number | bigint | null;
}

const CAP_FIELDS = ["max_turns", "time_budget_ms", "token_budget"];
const CHARGE_FIELDS = ["turns", "ms", "tokens"];

function ledger(value: number | bigint | null | undefined, what: string): bigint {
  if (value == null) return 0n; // absent, null or not given: nothing used
  return integer(value, what, 0n, LEDGER_MAX);
}

function cap(value: number | bigint | null | undefined, what: string): bigint | null {
  return value == null ? null : ledger(value, what);
}

function saturatingAdd(used: bigint, amount: bigint): bigint {
  const sum = used + amount;
  return sum > LEDGER_MAX ? LEDGER_MAX : sum;
}

export class Budget {
  readonly max_turns: bigint | null;
  readonly time_budget_ms: bigint | null;
  readonly token_budget: bigint | null;
  turns_used = 0n;
  ms_used = 0n;
  tokens_used = 0n;

  constructor(caps: BudgetCaps = {}) {
    checkFields(caps, CAP_FIELDS, "budget caps");
    this.max_turns = cap(caps.max_turns, "max_turns");
    this.time_budget_ms = cap(caps.time_budget_ms, "time_budget_ms");
    this.token_budget = cap(caps.token_budget, "token_budget");
  }

  /**
   * Add what was used and say where the budget stands. `warning` with the
   * first exhausted resource, in the order turns, time, tokens; `ok` while
   * every capped resource is below its cap. A charge of nothing reports the
   * current state, and an exhausted budget stays exhausted.
   */
  charge(charge: Charge = {}): Verdict {
    // Every amount is read before any is added: a refused charge changes nothing.
    checkFields(charge, CHARGE_FIELDS, "charge");
    const turns = ledger(charge.turns, "turns");
    const ms = ledger(charge.ms, "ms");
    const tokens = ledger(charge.tokens, "tokens");
    this.turns_used = saturatingAdd(this.turns_used, turns);
    this.ms_used = saturatingAdd(this.ms_used, ms);
    this.tokens_used = saturatingAdd(this.tokens_used, tokens);
    if (this.max_turns !== null && this.turns_used >= this.max_turns) {
      return verdict("warning", "budget_turns", `turns exhausted: ${this.turns_used} of ${this.max_turns} used`);
    }
    if (this.time_budget_ms !== null && this.ms_used >= this.time_budget_ms) {
      return verdict("warning", "budget_time", `time exhausted: ${this.ms_used} of ${this.time_budget_ms} ms used`);
    }
    if (this.token_budget !== null && this.tokens_used >= this.token_budget) {
      return verdict("warning", "budget_tokens", `tokens exhausted: ${this.tokens_used} of ${this.token_budget} used`);
    }
    return verdict(
      "ok",
      "budget_ok",
      `within budget: turns ${show(this.turns_used, this.max_turns)}, ms ${show(this.ms_used, this.time_budget_ms)}, tokens ${show(this.tokens_used, this.token_budget)}`,
    );
  }
}

File: go/budget.go
package haltrule

// Budget - a ledger of turns, time, and tokens against caps. Pure: it never
// reads a clock or counts a token; the caller charges what it measured.
//
// A budget is exhausted when the amount used reaches its cap, for all three
// resources alike. The ledger holds 64-bit integers and its additions saturate
// at 2^63 - 1, which is what int64 already is: the type carries the range the
// other ports check for. What is left to check is that an amount is not
// negative, and a negative one is refused, which is an error here.
//
// A nil cap is no cap and a nil amount is nothing used: null and not given are
// the same thing, as everywhere in this spec.

import (
	"errors"
	"fmt"
	"math"
)

const ledgerMax = math.MaxInt64

// BudgetCaps are the caps a budget is made with; each is nil for no cap.
type BudgetCaps struct {
	MaxTurns     *int64
	TimeBudgetMs *int64
	TokenBudget  *int64
}

// Charge is what was used since the last charge; each field is nil for nothing.
type Charge struct {
	Turns  *int64
	Ms     *int64
	Tokens *int64
}

// Budget is the ledger. Make one with NewBudget.
type Budget struct {
	caps       BudgetCaps
	TurnsUsed  int64
	MsUsed     int64
	TokensUsed int64
}

func amount(value *int64, what string) (int64, error) {
	if value == nil {
		return 0, nil
	}
	if *value < 0 {
		return 0, fmt.Errorf("%s must be within [0, 2^63 - 1], got %d", what, *value)
	}
	return *value, nil
}

// NewBudget refuses a cap that is not a non-negative integer.
func NewBudget(caps BudgetCaps) (*Budget, error) {
	for _, held := range []struct {
		value *int64
		what  string
	}{
		{caps.MaxTurns, "max_turns"},
		{caps.TimeBudgetMs, "time_budget_ms"},
		{caps.TokenBudget, "token_budget"},
	} {
		if _, err := amount(held.value, held.what); err != nil {
			return nil, err
		}
	}
	return &Budget{caps: caps}, nil
}

func saturatingAdd(used, added int64) int64 {
	if used > ledgerMax-added {
		return ledgerMax
	}
	return used + added
}

// Charge adds what was used and says where the budget stands: Warning naming
// the first exhausted resource, in the order turns, time, tokens, or OK while
// every capped resource is below its cap. A charge of nothing reports the
// current state, and an exhausted budget stays exhausted. A charge is refused
// whole: every amount is read before any is added, so a refused charge leaves
// the ledger exactly as it was.
func (b *Budget) Charge(charge Charge) (Verdict, error) {
	turns, turnsErr := amount(charge.Turns, "turns")
	ms, msErr := amount(charge.Ms, "ms")
	tokens, tokensErr := amount(charge.Tokens, "tokens")
	if err := errors.Join(turnsErr, msErr, tokensErr); err != nil {
		return Verdict{}, err
	}
	b.TurnsUsed = saturatingAdd(b.TurnsUsed, turns)
	b.MsUsed = saturatingAdd(b.MsUsed, ms)
	b.TokensUsed = saturatingAdd(b.TokensUsed, tokens)
	if b.caps.MaxTurns != nil && b.TurnsUsed >= *b.caps.MaxTurns {
		return verdict(Warning, "budget_turns",
			fmt.Sprintf("turns exhausted: %s used", showAmount(b.TurnsUsed, b.caps.MaxTurns))), nil
	}
	if b.caps.TimeBudgetMs != nil && b.MsUsed >= *b.caps.TimeBudgetMs {
		return verdict(Warning, "budget_time",
			fmt.Sprintf("time exhausted: %s ms used", showAmount(b.MsUsed, b.caps.TimeBudgetMs))), nil
	}
	if b.caps.TokenBudget != nil && b.TokensUsed >= *b.caps.TokenBudget {
		return verdict(Warning, "budget_tokens",
			fmt.Sprintf("tokens exhausted: %s used", showAmount(b.TokensUsed, b.caps.TokenBudget))), nil
	}
	return verdict(OK, "budget_ok", fmt.Sprintf("within budget: turns %s, ms %s, tokens %s",
		showAmount(b.TurnsUsed, b.caps.MaxTurns),
		showAmount(b.MsUsed, b.caps.TimeBudgetMs),
		showAmount(b.TokensUsed, b.caps.TokenBudget))), nil
}
